### PR TITLE
updatedocs

### PR DESCRIPTION
**Pull request by aviatesk**
Let's updates docs :)

- fix the not-working example of union expression matching
- update examples with static parameters and fully use the `where` syntax.

### CODE REVIEW

File: docs/src/utilities.md [hunk 1]
```diff
@@ -44,12 +44,33 @@ and returns `Dict(:name=>..., :args=>..., etc.)`. The definition can be rebuilt
...
 
 ```julia
-rtype = get(dict, :rtype, :Any)
-all_params = [get(dict, :params, [])..., get(dict, :whereparams, [])...]
-:(function $(dict[:name]){$(all_params...)}($(dict[:args]...);
-                                            $(dict[:kwargs]...))::$rtype
-      $(dict[:body])
-  end)
+rtype = get(dict, :rtype, nothing)
+wparams = get(dict, :whereparams, [])
```
Comment (MikeInnes) [inline docs/src/utilities.md]: This seems unnecessarily verbose. I guess the idea is to produce nicer code in some cases, but I think the simplest thing possible is better for teaching.

Comment (aviatesk) [reply docs/src/utilities.md]: okay, so I will try to mimic the previous documentation.

File: docs/src/pattern-matching.md [hunk 2]
```diff
@@ -97,18 +97,18 @@ which will match e.g. `struct Foo ...` but not `struct Foo{V} ...`
...
+@capture(ex, (f_(args__) = body_) | (function f_(args__) body_ end))
 ```
 
 will match both kinds of function syntax (though it's easier to use
 `shortdef` to normalise definitions). You can also do this within
 expressions, e.g.
 
 ```julia
-@capture(ex, (f_{T_}|f_)(args__) = body_)
+@capture(ex, (f_(args__) where {T_}) | (f_(args__)) = body_)
```
Comment (MikeInnes) [inline docs/src/pattern-matching.md]: It's nice to have the fix here, but this example no longer shows what it was intended to (an `|` pattern inside another pattern). Would be nice to come up with an example where we can show that.

Comment (aviatesk) [reply docs/src/pattern-matching.md]: umm I think
```jl
@capture(ex, (f_(args__) where {T_}) | (f_(args__)) = body_)
```
shows an union pattern in an expression (i.e. LHS expression in a shortform function def) ?

Comment (aviatesk) [reply docs/src/pattern-matching.md]: Or, how about this ?
```jl
@capture((Base.isconst|isconst)(args__))
```
to capture both calls (but kind of an nonsense example)

**Comment by aviatesk** [comment]
@MikeInnes could you give me your feedback ?

File: docs/src/pattern-matching.md (hunk 2 above)
Comment (o314) [reply docs/src/pattern-matching.md]: With `where` clause you will need `0:1` choice. That's will be very hard without epsilon transition to illustrate it via an `or` pattern. 

My advise keep the `func{params}(args)` which is still in use for constructor
(even considering https://github.com/JuliaLang/julia/issues/11310) to illustrate nested `or`.

Develop others forms notably with `where` as separated entries of a
```julia
@capture ex begin
pat1 => ...
pat2 => ...
end
```

**Comment by o314** [comment]
Other bonus point :
What is the purpose of `_str` regex matching in `match/*.jl` files ?
May be to handle non-standard string literals.
That's not documented anyway.

File: docs/src/utilities.md (hunk 1 above)
Comment (cstjean) [reply docs/src/utilities.md]: FWIW, as the splitdef author, I've found that I _always_ use `combinedef` to rebuild the definition, and I've been meaning to take out the "defintion can be rebuilt explicitly with" part entirely...

Also, `combinedef` should be exported...

Comment (aviatesk) [reply docs/src/utilities.md]: thanks for creating `splitdef`, @cstjean :)

I understand what you mean, still imho we still may want to show how to rebuild a definition expression from bare returned expressions, for a teaching purpose as @MikeInnes suggested above ?

File: docs/src/pattern-matching.md (hunk 2 above)
Comment (aviatesk) [reply docs/src/pattern-matching.md]: @MikeInnes

I still haven't came up with the good "an `|` pattern inside another pattern" example...
Could you give me any advice in order to finish up this PR, please ?

Comment (MikeInnes) [reply docs/src/pattern-matching.md]: Let's just stick to this, it's a little more complex but that's ok.

File: docs/src/utilities.md (hunk 1 above)
Comment (MikeInnes) [reply docs/src/utilities.md]: I think Julia will handle an empty `whereparams` list just fine. Probably worth testing to be sure, but if so we can avoid the branch.

Comment (aviatesk) [reply docs/src/utilities.md]: No, it doesn't work without the branching:
image

Comment (MikeInnes) [reply docs/src/utilities.md]: It works fine, it just doesn't print as nicely.

```julia
julia> :(f() where {} = 1)
:($(Expr(:where, :(f()))) = begin
          #= REPL[1]:1 =#
          1
      end)

julia> eval(ans)
f (generic function with 1 method)

julia> f()
1
```

Comment (aviatesk) [reply docs/src/utilities.md]: Ah yeah you're right. The example is simplified in the latest commit :)

**Comment by MikeInnes** [comment]
Thanks a lot @aviatesk!

**Pull request merged**